src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -341,6 +381,706 @@ CreateInsertSelectRouterPlan(Query *originalQuery,
+			lappend((*attributeEquivalanceClass)->equivalentAttributes,
+					attributeEqMember);
+	}
+	else if (rangeTableEntry->rtekind == RTE_SUBQUERY && !rangeTableEntry->inh)
+	{
+		Query *subquery = rangeTableEntry->subquery;
+		RelOptInfo *baseRelOptInfo = NULL;
+		TargetEntry *subqueryTargetEntry = NULL;
+
+		/* punt if it's a whole-row var rather than a plain column reference */
+		if (varToBeAdded->varattno == InvalidAttrNumber)
+		{
+			return;
+		}
+
+		baseRelOptInfo = find_base_rel(root, varToBeAdded->varno);
+
+		/* If the subquery hasn't been planned yet, we have to punt */
+		if (baseRelOptInfo->subroot == NULL)
+		{
+			return;
+		}
+
+		Assert(IsA(baseRelOptInfo->subroot, PlannerInfo));
+
+		subquery = baseRelOptInfo->subroot->parse;
+		Assert(IsA(subquery, Query));
+
+		/* Get the subquery output expression referenced by the upper Var */
+		subqueryTargetEntry = get_tle_by_resno(subquery->targetList,
+											   varToBeAdded->varattno);
+		if (subqueryTargetEntry == NULL || subqueryTargetEntry->resjunk)
+		{
+			ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+							errmsg("subquery %s does not have attribute %d",
+								   rangeTableEntry->eref->aliasname,
+								   varToBeAdded->varattno)));
+		}
+
+		if (!IsA(subqueryTargetEntry->expr, Var))
+		{
+			return;
+		}
+
+		varToBeAdded = (Var *) subqueryTargetEntry->expr;
+
+		/* we need to handle set operations separately */
+		if (subquery->setOperations)
+		{
+			SetOperationStmt *unionStatement =
+				(SetOperationStmt *) subquery->setOperations;
+
+			RangeTblRef *leftRangeTableReference = (RangeTblRef *) unionStatement->larg;
+			RangeTblRef *rightRangeTableReference = (RangeTblRef *) unionStatement->rarg;
+
+			varToBeAdded->varno = leftRangeTableReference->rtindex;
+			AddToAttributeEquivalenceClass(baseRelOptInfo->subroot, varToBeAdded,
+										   attributeEquivalanceClass);
+
+			varToBeAdded->varno = rightRangeTableReference->rtindex;
+			AddToAttributeEquivalenceClass(baseRelOptInfo->subroot, varToBeAdded,
+										   attributeEquivalanceClass);
+		}
+		else if (varToBeAdded && IsA(varToBeAdded, Var) && varToBeAdded->varlevelsup == 0)
+		{
+			AddToAttributeEquivalenceClass(baseRelOptInfo->subroot, varToBeAdded,
+										   attributeEquivalanceClass);
+		}
+	}
+}
+
+
+/*
+ * AttributeClassContainsAttributeClassMember returns true if it the input class member
+ * is already exists in the attributeEquivalenceClass. An equality is identified by the
+ * varattno and rteIdentity.
+ */
+static bool
+AttributeClassContainsAttributeClassMember(AttributeEquivalenceClassMember *inputMember,
+										   AttributeEquivalenceClass *
+										   attributeEquivalenceClass)
+{
+	ListCell *classCell = NULL;
+	foreach(classCell, attributeEquivalenceClass->equivalentAttributes)
+	{
+		AttributeEquivalenceClassMember *memberOfClass = lfirst(classCell);
+		if (memberOfClass->rteIdendity == inputMember->rteIdendity &&
```

Comment:
Yes, I agree that there are few places where the complexity is O(n^2) on the attribute number. But, now that I switched to adding only partition keys to the attribute equivalence, I started to think that it is OK given that we'd have very few attribute members.

Do you think should we find a better way of processing the equivalences, or are you OK with it?